services/dynamodb/src/test/java/software/amazon/awssdk/services/dynamodb/WaitersUserAgentTest.java

```diff
@@ -0,0 +1,106 @@
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.dynamodb.model.DescribeTableRequest;
+import software.amazon.awssdk.services.dynamodb.model.DescribeTableResponse;
+import software.amazon.awssdk.services.dynamodb.waiters.DynamoDbAsyncWaiter;
+import software.amazon.awssdk.services.dynamodb.waiters.DynamoDbWaiter;
+
+public class WaitersUserAgentTest {
+
+    @Rule
+    public WireMockRule mockServer = new WireMockRule(options().notifier(new ConsoleNotifier(true)).dynamicHttpsPort());
+
+    private DynamoDbClient dynamoDbClient;
+    private DynamoDbAsyncClient dynamoDbAsyncClient;
+    private ExecutionInterceptor interceptor;
+
+    @Before
+    public void setup() {
+        interceptor = Mockito.spy(AbstractExecutionInterceptor.class);
+
+        dynamoDbClient = DynamoDbClient.builder()
+                                       .credentialsProvider(StaticCredentialsProvider.create(AwsBasicCredentials.create("test",
+                                                                                                                        "test")))
+                                       .region(Region.US_WEST_2)
+                                       .endpointOverride(URI.create("http://localhost:" + mockServer.port()))
+                                       .overrideConfiguration(c -> c.addExecutionInterceptor(interceptor))
+                                       .build();
+
+        dynamoDbAsyncClient = DynamoDbAsyncClient.builder()
+                                                 .credentialsProvider(StaticCredentialsProvider.create(AwsBasicCredentials
+                                                                                                           .create("test",
+                                                                                                                   "test")))
+                                                 .region(Region.US_WEST_2)
+                                                 .endpointOverride(URI.create("http://localhost:" + mockServer.port()))
+                                                 .overrideConfiguration(c -> c.addExecutionInterceptor(interceptor))
+                                                 .build();
+    }
+
+    @Test
+    public void syncWaiters_shouldHaveWaitersUserAgent() {
+        stubFor(any(urlEqualTo("/")).willReturn(aResponse().withStatus(500)));
+
+        DynamoDbWaiter waiter = dynamoDbClient.waiter();
+        assertThatThrownBy(() -> waiter.waitUntilTableExists(DescribeTableRequest.builder().tableName("table").build())).isNotNull();
+
+        ArgumentCaptor<Context.BeforeTransmission> context = ArgumentCaptor.forClass(Context.BeforeTransmission.class);
+        Mockito.verify(interceptor, times(9)).beforeTransmission(context.capture(), Matchers.any());
+
+        assertTrue(context.getValue().httpRequest().headers().get("User-Agent").toString().contains("waiter"));
+    }
```

Comment:
How long do these tests take to run?

Reply:
The sync one takes around 7s, while the async one only takes 98ms, because the sync one does wait for the response. For the async one, the CompletableFuture is cancelled once the user agent is checked.

Reply:
Should we use mock clients instead? `codegen-generated-classes-test` seems like a good place for this test. https://github.com/aws/aws-sdk-java-v2/tree/waiters-development/test/codegen-generated-classes-test/src/test/java/software/amazon/awssdk/services/waiters

nit, you can just use wiremock's verify method rather than the interceptor
```
       verify(postRequestedFor(anyUrl()).withHeader("User-Agent", contains(...));
```

Reply:
You could also throw an exception from the interceptor if you want to cancel the calls altogether and just validate the request.

Reply:
@zoewangg I just tried testing with the mock clients and Mockito.verify(), and it took around 8 mins to get the result, not sure what caused that but it seems the WireMock test is much more efficient.